R/bootstrap.R

```diff
@@ -628,29 +691,37 @@ renv_bootstrap_library_root_impl <- function(project) {
 
 }
 
-renv_bootstrap_validate_version <- function(version) {
+renv_bootstrap_validate_version <- function(version, description = NULL) {
+  version_is_version <- grepl("[.-]", version)
+  description <- description %||% utils::packageDescription("renv")
 
-  loadedversion <- utils::packageDescription("renv", fields = "Version")
-  if (version == loadedversion)
-    return(TRUE)
+  if (version_is_version) {
+    loaded <- description$Version
+    if (identical(loaded, version)) {
+      return(TRUE)
+    }
+  } else {
+    loaded <- description$RemoteSha
+    if (startswith(loaded, version)) {
+      return(TRUE)
+    }
+  }
 
   # assume four-component versions are from GitHub;
   # three-component versions are from CRAN
-  components <- strsplit(loadedversion, "[.-]")[[1]]
-  remote <- if (length(components) == 4L)
-    paste("rstudio/renv", loadedversion, sep = "@")
-  else
-    paste("renv", loadedversion, sep = "@")
+  remote <- if (renv_bootstrap_version_is_dev(version)) {
+    paste("rstudio/renv", loaded, sep = "@")
+  } else {
+    paste("renv", loaded, sep = "@")
+  }
 
   fmt <- paste(
     "renv %1$s was loaded from project library, but this project is configured to use renv %2$s.",
-    "Use `renv::record(\"%3$s\")` to record renv %1$s in the lockfile.",
-    "Use `renv::restore(packages = \"renv\")` to install renv %2$s into the project library.",
+    "* Use `renv::record(\"%3$s\")` to record renv %1$s in the lockfile.",
+    "* Use `renv::restore(packages = \"renv\")` to install renv %2$s into the project library.",
     sep = "\n"
   )
-
-  msg <- sprintf(fmt, loadedversion, version, remote)
-  warning(msg, call. = FALSE)
+  catf(fmt, loaded, version, remote)
```

Comment:
I changed this to `catf()` because `warning` was messing up the new lines, it ensures that it's always displayed when you expected, it respects `renv.boostrap.quiet`, and I think we're generally moving towards more `cat()` usage (based on our experience in the tidyverse).

Reply:
Sounds good!